src/main/java/com/datadoghq/sketch/ddsketch/store/rescaling/RescalingStore.java

```diff
@@ -0,0 +1,125 @@
+package com.datadoghq.sketch.ddsketch.store.rescaling;
+
+import com.datadoghq.sketch.ddsketch.store.Bin;
+import com.datadoghq.sketch.ddsketch.store.Store;
+
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+public abstract class RescalingStore implements Store {
+
+    protected final Supplier<? extends Store> storeSupplier;
+
+    protected Store delegate;
+    protected int scalingFactor = 1;
+
+    public RescalingStore(Supplier<? extends Store> storeSupplier) {
+        this.storeSupplier = storeSupplier;
+        this.delegate = storeSupplier.get();
+    }
+
+    protected RescalingStore(Supplier<? extends Store> storeSupplier, Store delegate, int scalingFactor) {
+        this.storeSupplier = storeSupplier;
+        this.delegate = delegate;
+        this.scalingFactor = scalingFactor;
+    }
+
+    /**
+     * @return the rescaling factor > 1 if it should rescale
+     */
+    protected abstract int shouldRescale();
+
+    /**
+     * Merge contiguous buckets.
+     *
+     * @param rescalingFactor the factor to be compounded with the current one
+     */
+    protected final void rescale(int rescalingFactor) {
+        final Store newDelegate = storeSupplier.get();
+        final int newScalingFactor = scalingFactor * rescalingFactor;
+        delegate.getStream()
+            .map(bin -> new Bin(Math.floorDiv(bin.getIndex(), rescalingFactor), bin.getCount()))
+            .forEach(newDelegate::add);
```

Comment:
This is really succinct, but this sort of stuff really adds up. I think this could be done with a for loop and without allocation if `Store` had a method `forEach(BinAcceptor)`. where `BinAcceptor` is defined as follows:


````java
@FunctionalInterface
interface BinAcceptor {
    void accept(int index, double count);
}
```

Reply:
See #38